core/trino-main/src/main/java/io/trino/execution/scheduler/policy/SequentialExecutionSchedule.java

```diff
@@ -0,0 +1,606 @@
+                .map(stagesByFragmentId::get)
+                .forEach(this::startStage);
+    }
+
+    @Override
+    public Set<PipelinedStageExecution> getStagesToSchedule()
+    {
+        schedule();
+        return activeSources.stream()
+                .map(StageExecution::getPipelinedStageExecution)
+                .collect(toImmutableSet());
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        // dependency graph contains both running and not started fragments
+        return fragmentDependency.vertexSet().isEmpty();
+    }
+
+    @Override
+    public synchronized ListenableFuture<Void> getRescheduleFuture()
+    {
+        return rescheduleFuture;
+    }
+
+    @VisibleForTesting
+    void schedule()
+    {
+        removeCompletedStages();
+        unblockStages();
+    }
+
+    @VisibleForTesting
+    DirectedGraph<PlanFragmentId, FragmentsEdge> getFragmentDependency()
+    {
+        return fragmentDependency;
+    }
+
+    @VisibleForTesting
+    Set<StageExecution> getActiveSources()
+    {
+        return activeSources;
+    }
+
+    private void removeCompletedStages()
+    {
+        Set<StageExecution> completedStages = activeSources.stream()
+                .filter(this::isStageCompleted)
+                .collect(toImmutableSet());
+        // remove completed stages outside of Java stream to prevent concurrent modification
+        completedStages.forEach(this::removeCompletedStage);
+    }
+
+    private void removeCompletedStage(StageExecution stage)
+    {
+        // start all stages that depend on completed stage
+        PlanFragmentId fragmentId = stage.getFragment().getId();
+        fragmentDependency.outgoingEdgesOf(fragmentId).stream()
+                .map(FragmentsEdge::getTarget)
+                // filter stages that depend on completed stage only
+                .filter(dependentFragmentId -> fragmentDependency.inDegreeOf(dependentFragmentId) == 1)
+                .map(stagesByFragmentId::get)
+                .forEach(this::startStage);
+        fragmentDependency.removeVertex(fragmentId);
+        fragmentTopology.removeVertex(fragmentId);
+        activeSources.remove(stage);
+    }
+
+    private void unblockStages()
+    {
+        // find stages that are blocked on full task output buffer
+        Set<PlanFragmentId> blockedFragments = activeSources.stream()
+                .filter(StageExecution::isAnyTaskBlocked)
+                .map(stage -> stage.getFragment().getId())
+                .collect(toImmutableSet());
+        // start immediate downstream stages so that data can be consumed
+        blockedFragments.stream()
+                .flatMap(fragmentId -> fragmentTopology.outgoingEdgesOf(fragmentId).stream())
+                .map(FragmentsEdge::getTarget)
+                .map(stagesByFragmentId::get)
+                .forEach(this::startStage);
+    }
+
+    private void startStage(StageExecution stage)
```

Comment:
The class itself does not drives execution. Just provides ordering. So I think `selectForExecution`. Or `activate` or `makeActive` (for consistency with `activeStages` field) would be a better name.